src/main/java/org/jboss/modules/ModuleLoader.java

```diff
@@ -565,6 +583,21 @@ protected final void unloadModuleLocal(Module module) throws SecurityException {
         final String id = module.getIdentifier().toString();
         final FutureModule futureModule = moduleMap.get(id);
         if (futureModule != null && futureModule.module == module) {
+            final Set<String> aliases = module.aliases;
+            if (aliases != null) {
+                synchronized (aliases) {
+                    if (aliases.contains(id)) {
```

Comment:
I asked this once already, citing me and you and commenting inline:

> >  if (!module.aliases.contains(id)) condition necessary?
>
> The idea is alias module can be removed without invalidating aliased module.
> (In other words there is possibility alias module is unloaded at runtime).

I do not think unloading an alias module is possible. 

Let me explain it on an example: Let's say module `M1A` is an alias of module `M1` and there are no other modules in our universe. 

There will be two entries `M1A` and `M1` in `moduleMap`, but there will be just one instance of `org.jboss.modules.Module` in the JVM. The `name` of that `org.jboss.modules.Module` instance will be `M1`. Because there will be no `org.jboss.modules.Module` instance with name `M1A`, it is simply not be possible to call `unloadModuleLocal()` for that instance. 

The only possible call of `unloadModuleLocal()` is with `org.jboss.modules.Module` having name `M1` therefore `if (aliases.contains(id))` cannot be true at any reasonable circumstances.

That said, I think the whole `if (aliases.contains(id)) ... else ...` statment should be replaced with the content of the `else` branch.

> On the other hand when non alias module is unloaded, all its aliases should be invalidated too.
>
> >  Will module.aliases ever contain the id? Does it make any sense to register an alias for a module that is the same as its id?
>
>I don't think so.